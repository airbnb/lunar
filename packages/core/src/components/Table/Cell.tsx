import React from 'react';
import withStyles, { css, WithStylesProps } from '../../composers/withStyles';

export type Props = {
  /** Display in the center horizontally. */
  centerAlign?: boolean;
  /** Span multiple columns. */
  colSpan?: number;
  /** Display horizontally at the end. */
  endAlign?: boolean;
  /** Render as a table header. */
  header?: boolean;
  /** Display horizontally at the start. */
  startAlign?: boolean;
  /** Truncate text and display all on hover. */
  truncate?: boolean;
  /** Wrap text and white space. */
  wrap?: boolean;
};

/** An individual table cell. */
export class TableCell extends React.Component<Props & WithStylesProps> {
  static defaultProps = {
    centerAlign: false,
    endAlign: false,
    header: false,
    startAlign: false,
    truncate: false,
    wrap: false,
  };

  render() {
    const {
      centerAlign,
      children,
      header,
      startAlign,
      endAlign,
      styles,
      truncate,
      wrap,
      ...props
    } = this.props;
    const Tag = header ? 'th' : 'td';

    return (
      <Tag
        {...props}
        {...css(
          truncate && styles.cell_truncate,
          startAlign && styles.cell_left,
          centerAlign && styles.cell_center,
          endAlign && styles.cell_right,
          wrap && styles.cell_wrap,
        )}
      >
        {children}
      </Tag>
    );
  }
}

export default withStyles(() => ({
  cell_truncate: {
    overflow: 'hidden',
    whiteSpace: 'nowrap',
    textOverflow: 'ellipsis',

    ':hover': {
      maxWidth: 'none',
      overflow: 'inherit',
      whiteSpace: 'inherit',
      wordWrap: 'break-word',
    },
  },

  cell_left: {
    textAlign: 'left',
  },

  cell_center: {
    textAlign: 'center',
  },

  cell_right: {
    textAlign: 'right',
  },

  cell_wrap: {
    whiteSpace: 'normal',
    wordWrap: 'break-word',
  },
}))(TableCell);
