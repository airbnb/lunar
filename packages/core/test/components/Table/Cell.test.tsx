import React from 'react';
import { shallowWithStyles } from '@airbnb/lunar-test-utils';
import Cell from '../../../src/components/Table/Cell';

describe('<Cell />', () => {
  it('renders `td` by default', () => {
    const wrapper = shallowWithStyles(<Cell>Content</Cell>);

    expect(wrapper.type()).toBe('td');
  });

  it('renders `th` with `header`', () => {
    const wrapper = shallowWithStyles(<Cell header>Content</Cell>);

    expect(wrapper.type()).toBe('th');
  });

  it('renders left', () => {
    const wrapper = shallowWithStyles(<Cell startAlign>Left</Cell>);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders center', () => {
    const wrapper = shallowWithStyles(<Cell centerAlign>Center</Cell>);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders right', () => {
    const wrapper = shallowWithStyles(<Cell endAlign>Right</Cell>);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders truncate', () => {
    const wrapper = shallowWithStyles(<Cell truncate>Truncate</Cell>);

    expect(wrapper).toMatchSnapshot();
  });

  it('supports col span', () => {
    const wrapper = shallowWithStyles(<Cell colSpan={1}>Content</Cell>);

    expect(wrapper.prop('colSpan')).toBe(1);
  });
});
