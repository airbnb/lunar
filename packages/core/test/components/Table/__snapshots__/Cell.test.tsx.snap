// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Cell /> renders center 1`] = `
<td
  className="cell_center_12to336"
>
  Center
</td>
`;

exports[`<Cell /> renders left 1`] = `
<td
  className="cell_left_e4x16a"
>
  Left
</td>
`;

exports[`<Cell /> renders right 1`] = `
<td
  className="cell_right_140n9qh"
>
  Right
</td>
`;

exports[`<Cell /> renders truncate 1`] = `
<td
  className="cell_truncate_1sgzp64"
>
  Truncate
</td>
`;
